python/dgl/transform.py

```diff
@@ -143,6 +146,346 @@ def segmented_knn_graph(x, k, segs):
+            subgs[c_etype] = (F.cat([u, v], dim=0), F.cat([v, u], dim=0))
+
+        new_g = heterograph(subgs)
+    else:
+        subgs = {}
+        for c_etype in canonical_etypes:
+            if c_etype[0] != c_etype[2]:
+                u, v = g.all_edges(form='uv', order='eid', etype=c_etype)
+                subgs[c_etype] = (u, v)
+            else:
+                u, v = g.all_edges(form='uv', order='eid', etype=c_etype)
+                subgs[c_etype] = (F.cat([u, v], dim=0), F.cat([v, u], dim=0))
+
+        new_g = heterograph(subgs)
+
+    # handle features
+    if copy_ndata:
+        # for each ntype
+        for ntype in g.ntypes:
+            # for each data field
+            for k in g.nodes[ntype].data:
+                new_g.nodes[ntype].data[k] = g.nodes[ntype].data[k]
+
+    if copy_edata:
+        # for each etype
+        for c_etype in canonical_etypes:
+            if c_etype[0] != c_etype[2]:
+                # for each data field
+                for k in g.edges[c_etype].data:
+                    new_g.edges[c_etype].data[k] = g.edges[c_etype].data[k]
+            else:
+                for k in g.edges[c_etype].data:
+                    new_g.edges[c_etype].data[k] = \
+                        F.cat([g.edges[c_etype].data[k], g.edges[c_etype].data[k]], dim=0)
+    return new_g
+
+
+def joint_union(graph_list, copy_ndata=False, copy_edata=False):
+    r"""Return a joint union of the input graphs.
+
+    The new graph will include all the edges in the given
+    graphs. All the input graphs should have the same graph
+    structure and meta-information.
+
+    For heterographs with multiple edge types, we can
+    treat edges corresponding to each type as separate
+    graphs, and joint_union the graphs for each edge type.
+
+    Parameters
+    ----------
+    graphs : List of DGLGraph
+        The input graphs
+    copy_ndata: bool, optional
+        If True, the node features of the joint_unioned graph
+        are lazy-copied from the **FIRST** graph in the original
+        graph list. If False, the joint_unioned graph will not
+        have any node features.
+        (Default: False)
+    copy_edata: bool, optional
+        If True, the edge features of the joint_unioned graph
+        will be a concatenation of all the edge features from
+        the graphs in the original graph list. If False, the
+        joint_unioned graph will not have any edge features.
+        (Default: False)
+
+    Returns
+    -------
+    DGLGraph
+        The joint_unioned graph
+
+    Notes
+    -----
+    For heterographs with multiple edge types, all the input
+    graphs should have the same canonical edge types.
+
+    If ``copy_ndata`` is ``True``, same tensors is used for
+    the features of the original graph.
+    As a result, users should avoid performing in-place operations
+    on the features of the joint_unioned graph, which will corrupt
+    the features of the original graph as well.
```

Comment:
"on the features of the joint_unioned graph, which will corrupt the features of the original graph as well." -> "on the node features of the new graph to avoid feature corruption."